File: javascript/word-count/word-count.js
var wordCount = function(input) {
  var splitString = input.trim().split(/\s+/);

  var instances = splitString.reduce(function(hash, word) {
    if (!hash.hasOwnProperty(word)) { hash[word] = 0; }

    hash[word] += 1;
    return hash;
  }, {});

  return instances;
};

module.exports = wordCount;

File: clojure/word-count/src/word_count.clj
(ns word-count
  (:require [clojure.string :as str]))

(defn strip-punctuation
  "Removes punctuation"
  [word]
  (re-find (re-pattern "[a-z0-9][a-z]*") word))

(defn categorize
  "Counts the words"
  [hash word]
  (update-in hash [word] (fnil inc 0)))

(defn word-count
  "Given a phrase it counts the occurrences of each word in that phrase"
  [phrase]
     (reduce categorize (hash-map)
      (->> (str/split phrase #" ")
           (map str/lower-case)
           (map strip-punctuation)
           (remove nil?))))
